Prose/ProseLib/Sources/ConversationFeature/Chat/Chat.swift

```diff
@@ -254,20 +262,64 @@ struct ChatView: NSViewRepresentable {
     signposter.endInterval(#function, interval)
   }
 
-  func updateMessages(to messages: [ProseCoreViewsMessage], in webView: WKWebView) {
-    logger.trace("Updating \(Self.self): \(messages.count, privacy: .public) messages")
-
+  func updateMessages(
+    to messages: [ProseCoreViewsMessage],
+    in webView: WKWebView,
+    coordinator: Coordinator
+  ) {
     let interval = signposter.beginInterval(#function, id: self.signpostID)
 
-    let jsonData: Data = try! JSONEncoder().encode(messages)
-    let json = String(data: jsonData, encoding: .utf8) ?? "[]"
+    let messages = IdentifiedArrayOf<ProseCoreViewsMessage>(uniqueElements: messages)
+    let oldMessages: IdentifiedArrayOf<ProseCoreViewsMessage> = coordinator.messages
+    defer { coordinator.messages = messages }
+
+    let ids = OrderedSet<Message.ID>(messages.map(\.id))
+    let oldIds: OrderedSet<Message.ID> = coordinator.messages.ids
+
+    let insertedIds: OrderedSet<Message.ID> = ids.subtracting(oldIds)
+    let removedIds: OrderedSet<Message.ID> = oldIds.subtracting(ids)
```

Comment:
Please extract that diffing logic and add a few rudimentary tests at least. Doesn't need to be generic or anything and can live in the (yet to be introduced) ConversationFeatureTests.

Reply:
I made it generic to make testing easier